### ISSUE TITLE
How to use with animations?

### ISSUE THREAD

**Issue by kylecannon** [getsaf/shallow-render#59]
First off, thank you so much for the library. I am having an issue when I have animations triggers though with using this library. Example:

```typescript
@Component({
    selector: 'app-applications-grid',
    templateUrl: './applications-grid.component.html',
    styleUrls: ['./applications-grid.component.scss'],
    changeDetection: ChangeDetectionStrategy.OnPush,
    animations: [
        trigger('applicationAnimation', [
            transition('* => *', [
                query(
                    ':enter',
                    [
                        style({ opacity: 0, transform: 'translateY(-100px)' }),
                        stagger(50, [animate('400ms ease-out', style({ opacity: 1, transform: 'none' }))]),
                    ],
                    { optional: true }
                ),
            ]),
        ]),
    ],
})
export class ApplicationsGridComponent {
    @Input() applications: ApplicationDetail[];

    trackByFn(idx, { name }: ApplicationDetail) {
        return name;
    }
}
```

with a template of
```html
<section class="application-grid" [@applicationAnimation]="applications?.length">
    <app-applications-grid-item *ngFor="let application of applications; trackBy: trackByFn" [item]="application"></app-applications-grid-item>
</section>
```

This error is thrown:

```
HeadlessChrome 0.0.0 (Linux 0.0.0) ApplicationsGridComponent should create FAILED
	Error: Found the synthetic property @applicationAnimation. Please include either "BrowserAnimationsModule" or "NoopAnimationsModule" in your application.
	    at checkNoSyntheticProp node_modules/@angular/platform-browser/fesm5/platform-browser.js:1296:1)
	    at EmulatedEncapsulationDomRenderer2.push../node_modules/@angular/platform-browser/fesm5/platform-browser.js.DefaultDomRenderer2.setProperty node_modules/@angular/platform-browser/fesm5/platform-browser.js:1280:1)
	    at DebugRenderer2.push../node_modules/@angular/core/fesm5/core.js.DebugRenderer2.setProperty node_modules/@angular/core/fesm5/core.js:11207:1)
	    at setElementProperty node_modules/@angular/core/fesm5/core.js:7966:1)
	    at checkAndUpdateElementValue node_modules/@angular/core/fesm5/core.js:7917:1)
	    at checkAndUpdateElementInline node_modules/@angular/core/fesm5/core.js:7864:1)
	    at checkAndUpdateNodeInline node_modules/@angular/core/fesm5/core.js:10205:1)
	    at checkAndUpdateNode node_modules/@angular/core/fesm5/core.js:10171:1)
	    at debugCheckAndUpdateNode node_modules/@angular/core/fesm5/core.js:10804:1)
	    at debugCheckRenderNodeFn node_modules/@angular/core/fesm5/core.js:10790:1)
HeadlessChrome 0.0.0 (Linux 0.0.0) ApplicationsGridComponent should create FAILED
	Error: Found the synthetic property @applicationAnimation. Please include either "BrowserAnimationsModule" or "NoopAnimationsModule" in your application.
	    at checkNoSyntheticProp node_modules/@angular/platform-browser/fesm5/platform-browser.js:1296:1)
	    at EmulatedEncapsulationDomRenderer2.push../node_modules/@angular/platform-browser/fesm5/platform-browser.js.DefaultDomRenderer2.setProperty node_modules/@angular/platform-browser/fesm5/platform-browser.js:1280:1)
	    at DebugRenderer2.push../node_modules/@angular/core/fesm5/core.js.DebugRenderer2.setProperty node_modules/@angular/core/fesm5/core.js:11207:1)
	    at setElementProperty node_modules/@angular/core/fesm5/core.js:7966:1)
	    at checkAndUpdateElementValue node_modules/@angular/core/fesm5/core.js:7917:1)
	    at checkAndUpdateElementInline node_modules/@angular/core/fesm5/core.js:7864:1)
	    at checkAndUpdateNodeInline node_modules/@angular/core/fesm5/core.js:10205:1)
	    at checkAndUpdateNode node_modules/@angular/core/fesm5/core.js:10171:1)
	    at debugCheckAndUpdateNode node_modules/@angular/core/fesm5/core.js:10804:1)
```

If I create a testing module that imports the `NoopAnimationsModule`,

```typescript
@NgModule({
    imports: [NoopAnimationsModule, ApplicationsModule],
    exports: [ApplicationsModule],
})
export class ApplicationsTestModule {
}

describe('ApplicationsGridComponent', () => {
    let shallow: Shallow<ApplicationsGridComponent>;

    beforeEach(() => {
        shallow = new Shallow(ApplicationsGridComponent, ApplicationsTestModule);
    });

    const applications = mockApps;

    it('should create', async () => {
        const res = await shallow.render(
            `<app-applications-grid [applications]="applications"></app-applications-grid>`,
            { bind: { applications } },
        );
    });
});
```

 the test fails with the following error:

```
HeadlessChrome 0.0.0 (Linux 0.0.0) ApplicationsGridComponent should create FAILED
	TypeError: this.delegate.createRenderer is not a function
	    at DebugRendererFactory2.push../node_modules/@angular/core/fesm5/core.js.DebugRendererFactory2.createRenderer node_modules/@angular/core/fesm5/core.js:11067:1)
	    at createRootData node_modules/@angular/core/fesm5/core.js:10538:1)
	    at Object.debugCreateRootView [as createRootView] node_modules/@angular/core/fesm5/core.js:10531:1)
	    at ComponentFactory_.push../node_modules/@angular/core/fesm5/core.js.ComponentFactory_.create node_modules/@angular/core/fesm5/core.js:8361:1)
	    at initComponent node_modules/@angular/core/fesm5/testing.js:1163:1)
	    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke node_modules/zone.js/dist/zone.js:388:1)
	    at ProxyZoneSpec.push../node_modules/zone.js/dist/zone-testing.js.ProxyZoneSpec.onInvoke node_modules/zone.js/dist/zone-testing.js:288:1)
	    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke node_modules/zone.js/dist/zone.js:387:1)
	    at Object.onInvoke node_modules/@angular/core/fesm5/core.js:3671:1)
	    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke node_modules/zone.js/dist/zone.js:38HeadlessChrome 0.0.0 (Linux 0.0.0): Executed 1 of 125 (1 FAILED) (skipped 124) ERROR (0.096 secs / 0.072 secs)
```

Any ideas?

**Comment by getsaf** [comment]
Thanks for the positive feedback! You're close here but instead of creating a separate test module, you'll use your actual Angular module and ["replace"](https://github.com/getsaf/shallow-render/wiki/Mocking#replace-a-module-with-a-test-module) `BrowserAnimationsModule` with `NoopAnimationsModule` with Shallow.

```typescript
describe('ApplicationsGridComponent', () => {
    let shallow: Shallow<ApplicationsGridComponent>;

    beforeEach(() => {
        shallow = new Shallow(ApplicationsGridComponent, ApplicationsModule)
          .replaceModule(BrowserAnimationsModule, NoopAnimationsModule);
    });

    const applications = mockApps;

    it('should create', async () => {
        const res = await shallow.render(
            `<app-applications-grid [applications]="applications"></app-applications-grid>`,
            { bind: { applications } },
        );
    });
});
```

If you want to do this for all tests (recommended), you can add this line in your test shim to replace that module globally:

```typescript
Shallow.alwaysReplaceModule(BrowserAnimationsModule, NoopAnimationsModule);
```

Let me know if that works out for you!

**Comment by kylecannon** [comment]
I just tried that and it still doesn't work. The ApplicationsModule isn't the root module so it never actually imports the BrowserAnimationsModule since AppModule imports it.

```typescript
fdescribe('ApplicationsGridComponent', () => {
    let shallow: Shallow<ApplicationsGridComponent>;

    beforeEach(() => {
        shallow = new Shallow(ApplicationsGridComponent, ApplicationsModule)
            .replaceModule(BrowserAnimationsModule, NoopAnimationsModule);
    });

    const applications = mockApps;

    it('should create', async () => {
        const res = await shallow.render(
            `<app-applications-grid [applications]="applications"></app-applications-grid>`,
            { bind: { applications } },
        );
    });
});

````

returns:

```
	Error: Found the synthetic property @applicationAnimation. Please include either "BrowserAnimationsModule" or "NoopAnimationsModule" in your application.
	    at checkNoSyntheticProp node_modules/@angular/platform-browser/fesm5/platform-browser.js:1296:1)
	    at EmulatedEncapsulationDomRenderer2.push../node_modules/@angular/platform-browser/fesm5/platform-browser.js.DefaultDomRenderer2.setProperty node_modules/@angular/platform-browser/fesm5/platform-browser.js:1280:1)
	    at DebugRenderer2.push../node_modules/@angular/core/fesm5/core.js.DebugRenderer2.setProperty node_modules/@angular/core/fesm5/core.js:11207:1)
	    at setElementProperty node_modules/@angular/core/fesm5/core.js:7966:1)
	    at checkAndUpdateElementValue node_modules/@angular/core/fesm5/core.js:7917:1)
	    at checkAndUpdateElementInline node_modules/@angular/core/fesm5/core.js:7864:1)
	    at checkAndUpdateNodeInline node_modules/@angular/core/fesm5/core.js:10205:1)
	    at checkAndUpdateNode node_modules/@angular/core/fesm5/core.js:10171:1)
	    at debugCheckAndUpdateNode node_modules/@angular/core/fesm5/core.js:10804:1)
	    at debugCheckRenderNodeFn node_modules/@angular/core/fesm5/core.js:10790:1)
HeadlessChrome 0.0.0 (Linux 0.0.0) ApplicationsGridComponent should create FAILED
	Error: Found the synthetic property @applicationAnimation. Please include either "BrowserAnimationsModule" or "NoopAnimationsModule" in your application.
	    at checkNoSyntheticProp node_modules/@angular/platform-browser/fesm5/platform-browser.js:1296:1)
	    at EmulatedEncapsulationDomRenderer2.push../node_modules/@angular/platform-browser/fesm5/platform-browser.js.DefaultDomRenderer2.setProperty node_modules/@angular/platform-browser/fesm5/platform-browser.js:1280:1)
	    at DebugRenderer2.push../node_modules/@angular/core/fesm5/core.js.DebugRenderer2.setProperty node_modules/@angular/core/fesm5/core.js:11207:1)
	    at setElementProperty node_modules/@angular/core/fesm5/core.js:7966:1)
	    at checkAndUpdateElementValue node_modules/@angular/core/fesm5/core.js:7917:1)
	    at checkAndUpdateElementInline node_modules/@angular/core/fesm5/core.js:7864:1)
	    at checkAndUpdateNodeInline node_modules/@angular/core/fesm5/core.js:10205:1)
	    at checkAndUpdateNode node_modules/@angular/core/fesm5/core.js:10171:1)
	    at debugCheckAndUpdateNode node_modules/@angular/core/fesm5/core.js:10804:1)

```

**Comment by getsaf** [comment]
I believe your component's module should also import the BrowserAnimationModule. It's ok if it's also imported from the root but my guess is that you don't really need it in the root. My general rule of thumb is that the module for a component (or components) will have everything the children need. The root module (app module in your case) can generally only import singleton modules (eg: `forRoot` modules) and your component modules. Generally, your root (app) module does not concern itself with importing dependencies of child components (think single-responsibility-principle here).

Hope that doesn't come across as "preachy", just wanted to put out the thought process behind self-sufficient component modules.

Long story short, you should be ok to import the AnimationModule into your component's module and potentially remove it from the root (app) module.

**Comment by getsaf** [comment]
Hey @kylecannon, did this take care of your issue? Looking to close this issue if you're all good.

**Comment by kylecannon** [comment]
I would also like to add that the main entry page on my application immediately needs animations so keeping at the root is not a bad thing for my instance. I do see your point but the other issue with importing the animations module on my lazy loaded route is it also adds an additional 30kb of overhead for that particular code-split ngmodule. We are trying to keep our bundle sizes down as much as possible.

**Comment by getsaf** [comment]
I had not considered the bundle size like you mentioned. That alone may merit allowing shallow to add imports to the test module. I'll see what I can do for the next release.

Thinking something simple like:
```typescript
new Shallow(MyComponent, MyModule)
  .import(NoopAnimationsModule)
  .dontMock(NoopAnimationsModule);
```

And for globally importing a thing:
```typescript
Shallow.alwaysImport(FooModule);
```

We use ngrx where I work too, I haven't had any issues with testing it. Do you have an example of the issue? Maybe a stackblitz?

**Comment by getsaf** [comment]
Aww man, sorry, I just checked out our code and noticed that we are using ngrx-store not ngxs-store.
There was a guy that I talked to a while back about testing with ngxs #36

**Comment by kylecannon** [comment]
We use ngrx as well just can’t get it to work nicely or easily for that manner. Do you have examples of ngrx? Specifically with store feature modules.

**Comment by getsaf** [comment]
I don't have any examples, the team who wrote that code in my company's project currently has 0 test coverage. I'd have to play around with it on the [shallow-render stackblitz](https://stackblitz.com/github/getsaf/shallow-render-stackblitz) to see if I can get something demo-able. From what I remember, I *think* it'd just be a matter of mocking the `Store` provider to return an observable that contains all the things I'd want from the store. Without messing with it myself, I'm probably over-simplifying things though.

I'm going to be on vacation next week so it might be a little bit before I get to it. I will get to it though if you don't find a solution first.

**Comment by getsaf** [comment]
I got something decent here... [check it](https://stackblitz.com/edit/ngrx-store-shallow-render?file=examples/with-ngrx-store.spec.ts)

*Take this with a grain of salt, I think this is correct but I've only written tests against this contrived component in this example stackblitz so I don't know how it will pan out for your components unless they are using the same patterns*

Basically just mocked the store outright but had to pull the root providers into shallow similar to how the [docs suggest pulling in `forRoot` modules](https://github.com/getsaf/shallow-render/wiki/Mocking#global-providers-with-alwaysprovide).

#### With a component like this
```typescript
@Component({
  selector: 'my-component',
  template: '<div>{{count}}-{{resetCount}}</div>',
})
class MyComponent {
  count: number;
  resetCount: number;

  constructor(store: Store<CountState>) {
    store.subscribe(state => {
      this.count = state.count;
      this.resetCount = state.resetCount;
    })
  }
}
```

```typescript
describe('MyComponent', () => {
  let shallow: Shallow<MyComponent>;
  let mockStoreSubject: BehaviorSubject<CountState>;

  beforeEach(() => {
    mockStoreSubject = new BehaviorSubject<CountState>({
      count: 0,
      resetCount: 0,
    });

    shallow = new Shallow(MyComponent, MyModule)
      .provide(StoreModule.forRoot({count: counterReducer}).providers)
      .mock(
        Store,
        { subscribe: (...args) => mockStoreSubject.subscribe(...args) }
      );
  });

  it('displays count and reset count', async () => {
    mockStoreSubject.next({
      count: 33,
      resetCount: 12,
    });
    const { find } = await shallow.render();

    expect(find('div').nativeElement.innerText).toBe('33-12');
  });

  it('displays updated count and reset count', async () => {
    const { fixture, find } = await shallow.render();
    mockStoreSubject.next({
      count: 33,
      resetCount: 12,
    });
    fixture.detectChanges();

    expect(find('div').nativeElement.innerText).toBe('33-12');
  });
});
```

Hope this helps.

**Comment by getsaf** [comment]
Closing this issue, I think the Animations issue is solved here.

**Comment by kylecannon** [comment]
Hey @getsaf,

Are you still planning on the change you proposed here: https://github.com/getsaf/shallow-render/issues/59#issuecomment-418721470

**Comment by getsaf** [comment]
I haven't but it's not off the table. I plan to do some work and cut a new release this weekend. I'll try to get it done.

**Comment by getsaf** [comment]
Hey @kylecannon not sure if you saw, I published a new version that includes the feature you requested.

`Shallow.alwaysImport` and `shallow.import` are in #68 (see v6.2.x)

**Comment by kylecannon** [comment]
Thanks @getsaf I'll take a look as soon as I get a minute.

**Comment by lubek1983** [comment]
> This mocked store doesn't have `dispatch`. At my current client I wrote a mock store that lets you `next` but also `forward` from another observable, which is great for marble testing.
> 
> ```ts
> export interface MockStore<T> extends Observable<T> {
>   forward(source: Observable<T>): void;
>   next(state: T): void;
>   dispatch(): void;
> }
> 
> export const makeAbstractMockStore = <T>() => {
>   const subject = new ReplaySubject<Observable<T>>(1);
>   const observable = subject.pipe(
>     switchAll(),
>     shareReplay(1),
>   ) as MockStore<T>;
> 
>   observable.forward = (source: Observable<T>) => {
>     subject.next(source);
>   };
> 
>   observable.next = (state: T) => {
>     subject.next(of(state));
>   };
> 
>   observable.dispatch = () => {};
> 
>   return observable;
> };
> ```

could you give whole example how to use it ?

**Comment by ohjames** [comment]
@lubek1983 If it isn't already obvious then you probably don't need it.

**Comment by lubek1983** [comment]
> @lubek1983 If it isn't already obvious then you probably don't need it.

well i still dont fell how to use / mock things via shallow-render

i tried to do smth like this with your function

 `    beforeEach(() => {
        shallow = new Shallow(AssetExpandableComponent, SharedModule)
            .provide(StoreModule.forRoot({ data: confirmsReducer }).providers)
            .mock(
                Store, makeAbstractMockStore(),
            );
    });`
and then im trying to build this kind of test that will check that dispatch have been called when nested component was clicked
`    it('should contain sps button with click method', async () => {
        const { find } = await shallow
            .render();

        const button = find('sps-button-icon-only')[0];
        button.nativeElement
            .click();
        expect()
            .toHaveBeenCalled();

    });`
but still dont know how to do this

**Comment by ohjames** [comment]
@lubek1983 I'd do the testing via jest-marbles or jasmine-marbles if it was me. But for a simple use case, you should be storing a reference to the mock store... otherwise you can't push mock data through it. Although this kind of support is proper best handled through stackoverflow than a GitHub issue.
